snarky_universe/hash.ml

```diff
@@ -0,0 +1,128 @@
+type alpha = [`_5 | `_11]
+
+let alpha_of_curve = function Curve.Bn128 -> `_5
+
+let params_of_curve : Curve.t -> string Sponge.Params.t = function
+  | Bn128 ->
+      Sponge.Params.bn128
+
+module Poseidion_rounds = struct
+  module type S = sig
+    val rounds_full : int
+
+    val rounds_partial : int
+  end
+
+  type t = (module S)
+
+  let alpha_5 : t =
+    ( module struct
+      let rounds_full = 8
+
+      let rounds_partial = 83
+    end )
+
+  let alpha_11 : t =
+    ( module struct
+      let rounds_full = 8
+
+      let rounds_partial = 33
+    end )
+
+  let of_alpha = function `_5 -> alpha_5 | `_11 -> alpha_11
+end
+
+module Make
+    (Impl : Snarky.Snark_intf.Run) (C : sig
+        val curve : Curve.t
+    end) =
+struct
+  open Core_kernel
+  open Impl
+
+  module Inputs = struct
+    include (val Poseidion_rounds.of_alpha (alpha_of_curve C.curve))
+
+    module Field = struct
+      (* The linear combinations involved in computing Poseidon do not involve very many
+    variables, but if they are represented as arithmetic expressions (that is, "Cvars"
+    which is what Field.t is under the hood) the expressions grow exponentially in
+    in the number of rounds. Thus, we compute with Field elements represented by
```

Comment:
Perhaps not immediately, but we should add something to deal with this better in general.